src/createSlice.ts

```diff
@@ -265,7 +350,86 @@ export function createSlice<
   return {
     name,
     reducer,
-    actions: actionCreators as any,
-    caseReducers: sliceCaseReducersByName as any
+    actions: context.actionCreators as any,
+    caseReducers: context.sliceCaseReducersByName as any
+  }
+}
+
+// --- internal functions & types ---
+
+interface ReducerHandlingContext<State> {
+  sliceCaseReducersByName: Record<string, CaseReducer<State, any>>
+  sliceCaseReducersByType: Record<string, CaseReducer<State, any>>
+  actionCreators: Record<string, Function>
+}
+
+interface ReducerDetails {
+  reducerName: string
+  type: string
+}
+
+function getReducerCreators<State>(): ReducerCreators<State> {
+  return {
+    asyncThunk(payloadCreator, config) {
+      return {
+        [reducerDefinitionType]: ReducerType.asyncThunk,
+        ...config,
+        payloadCreator
+      }
+    }
+  }
+}
+
+function handleNormalReducerDefinition<State>(
+  { type, reducerName }: ReducerDetails,
+  maybeReducerWithPrepare:
+    | CaseReducer<State, { payload: any; type: string }>
+    | CaseReducerWithPrepare<State, PayloadAction<any, string, any, any>>,
+  context: ReducerHandlingContext<State>
+) {
+  let caseReducer: CaseReducer<State, any>
+  let prepareCallback: PrepareAction<any> | undefined
+  if ('reducer' in maybeReducerWithPrepare) {
+    caseReducer = maybeReducerWithPrepare.reducer
+    prepareCallback = maybeReducerWithPrepare.prepare
+  } else {
+    caseReducer = maybeReducerWithPrepare
+  }
+  context.sliceCaseReducersByName[reducerName] = caseReducer
+  context.sliceCaseReducersByType[type] = caseReducer
+  context.actionCreators[reducerName] = prepareCallback
+    ? createAction(type, prepareCallback)
+    : createAction(type)
+}
+
+function isAsyncThunkSliceReducerDefinition<State>(
+  reducerDefinition: any
+): reducerDefinition is AsyncThunkSliceReducerDefinition<State, any, any, any> {
+  return reducerDefinition[reducerDefinitionType] === ReducerType.asyncThunk
+}
+
+function handleThunkCaseReducerDefinition<State>(
+  { type, reducerName }: ReducerDetails,
+  reducerDefinition: AsyncThunkSliceReducerDefinition<State, any, any, any>,
+  context: ReducerHandlingContext<State>
+) {
+  const {
+    payloadCreator,
+    fulfilledReducer,
+    pendingReducer,
+    rejectedReducer,
+    options
+  } = reducerDefinition
+  const thunk = createAsyncThunk(type, payloadCreator, options)
+  context.actionCreators[reducerName] = thunk
+  // TODO: do we want to support `sliceCaseReducersByName` here? How would it look?
```

Comment:
this.

Reply:
Currently we are exposing the reducers as `slice.reducers[reducerName]` so they can be re-used from within other reducers. 
As this would potentially add up to three reducers that would also all be more like `extraReducers`, we are at the moment not exposing these extra thunk reducers. The question is if we want to keep it that way.

Reply:
Hmm.  Good question. I'm guessing it would be hard to name them, say, `myThunkPending` or something and have that typed correctly?

Reply:
Hm. TS being TS we can't do string concatenation. `myThunk.pending` would be possible though.

Reply:
done